### PR TITLE
Proxy Protocol support for Proxy SSH listener

### LINKED ISSUE #7924: Proxy Protocol & Proxy SSH listener issues
Customer has reported Teleport is not honoring Proxy Protocol V1 when emitting audit events for when a user creates a session.

---

The Proxy SSH listener (3023) does not support Proxy Protocol. When connecting with PROXY V1, the client's real address is not captured, and when connecting with PROXY V2, the binary format results in the server SSH conn failing to instantiate, leading to the connection being dropped.

As many customers do not make use of the ALPN multiplexing, we should aim to support Proxy protocol on this port. We can do this by leveraging the existing multiplexer code, and using that to correct the remote address. This will resolve the connection dropping when PROXY V2 is used, and ensure the correct IP address is captured on both protocols.

### PR DESCRIPTION
**Pull request by strideynet**
Closes #7924

Tested with:
- AWS NLB, PROXY protocol v2
- AWS CLB, PROXY protocol v1

This PR also modifies the Multiplexer to correctly recognise incoming SSH connections with the Teleport `ProxyHelloSignature` header.

### CODE REVIEW

**Comment by strideynet** [comment]
Tested with AWS NLB, PROXY V2 enabled, no longer crashes and now shows the correct IP.

Going to write some test cases to cover this, and it looks like its causing some other tests to fail.

**Comment by strideynet** [comment]
Test failures: https://gist.github.com/strideynet/4f413469d474fd71c032ea9c1d795018

**Comment by strideynet** [comment]
I've been able to reproduce this by running the package tests locally with shuffling enabled. This confirms to me that there must be some shared configuration being modified by another test. Now that I'm able to reproduce this, it becomes easier to work out what changes are required to fix this.

File: lib/sshutils/server.go [hunk 1]
```diff
@@ -417,6 +418,14 @@ func (s *Server) HandleConnection(conn net.Conn) {
 	wrappedConn := wrapConnection(wconn, s.log)
 	sconn, chans, reqs, err := ssh.NewServerConn(wrappedConn, &s.cfg)
 	if err != nil {
+		// Ignore EOF as these are triggered by loadbalancer health checks
+		if !errors.Is(err, io.EOF) {
+			s.log.
```
Comment (strideynet) [inline lib/sshutils/server.go:423]: I've introduced a log message here. Sinking this error previously was what meant that it was difficult to work out what was up with the incoming connections.

File: lib/multiplexer/multiplexer_test.go [hunk 2]
```diff
@@ -721,6 +731,70 @@ func TestMux(t *testing.T) {
...
+		// detected as SSH connections.
+		t.Parallel()
+		listener, err := net.Listen("tcp", "127.0.0.1:0")
+		require.Nil(t, err)
+
+		mux, err := New(Config{
+			Listener:            listener,
+			EnableProxyProtocol: true,
+		})
+		require.Nil(t, err)
```
Comment (jakule) [inline lib/multiplexer/multiplexer_test.go:746]:
```suggestion
		require.NoError(t, err)
```

File: lib/multiplexer/multiplexer_test.go [hunk 3]
```diff
@@ -721,6 +731,70 @@ func TestMux(t *testing.T) {
 		require.NoError(t, err)
 		require.Equal(t, "db listener", string(dbBytes))
 	})
+
+	t.Run("SSHProxyHelloSignature", func(t *testing.T) {
+		// Ensures SSH connections fronted with the ProxyHelloSignature are
+		// detected as SSH connections.
+		t.Parallel()
+		listener, err := net.Listen("tcp", "127.0.0.1:0")
+		require.Nil(t, err)
```
Comment (jakule) [inline lib/multiplexer/multiplexer_test.go:740]:
```suggestion
		require.NoError(t, err)
```

File: lib/multiplexer/multiplexer_test.go [hunk 4]
```diff
@@ -721,6 +731,70 @@ func TestMux(t *testing.T) {
...
+		})
+		require.Nil(t, err)
+		go mux.Serve()
+		defer mux.Close()
+
+		called := false
+		sshHandler := sshutils.NewChanHandlerFunc(func(_ context.Context, _ *sshutils.ConnectionContext, nch ssh.NewChannel) {
+			called = true
+			err := nch.Reject(ssh.Prohibited, "nothing to see here")
+			require.Nil(t, err)
```
Comment (jakule) [inline lib/multiplexer/multiplexer_test.go:754]:
```suggestion
			require.NoError(t, err)
```

File: lib/multiplexer/multiplexer_test.go [hunk 5]
```diff
@@ -721,6 +731,70 @@ func TestMux(t *testing.T) {
...
+			Timeout:         time.Second,
+			HostKeyCallback: ssh.FixedHostKey(signer.PublicKey()),
+		})
+		require.NoError(t, err)
+		clt := ssh.NewClient(c, chans, reqs)
+		defer clt.Close()
+
+		// call new session to initiate opening new channel
+		_, err = clt.NewSession()
+		require.NotNil(t, err)
```
Comment (jakule) [inline lib/multiplexer/multiplexer_test.go:790]:
```suggestion
		require.NoError(t, err)
```

File: lib/multiplexer/multiplexer_test.go [hunk 6]
```diff
@@ -721,6 +731,70 @@ func TestMux(t *testing.T) {
...
+		// make sure the channel handler was called OK
+		require.Equal(t, called, true)
```
Comment (jakule) [inline lib/multiplexer/multiplexer_test.go:792]:
```suggestion
		require.True(t, called)
```

File: lib/multiplexer/multiplexer_test.go [hunk 7]
```diff
@@ -721,6 +731,70 @@ func TestMux(t *testing.T) {
...
+		})
+
+		srv, err := sshutils.NewServer(
+			"test",
+			utils.NetAddr{AddrNetwork: "tcp", Addr: "localhost:0"},
+			sshHandler,
+			[]ssh.Signer{signer},
+			sshutils.AuthMethods{Password: pass("abc123")},
+		)
+		require.Nil(t, err)
```
Comment (jakule) [inline lib/multiplexer/multiplexer_test.go:764]:
```suggestion
		require.NoError(t, err)
```

**Review by jakule — APPROVED** [review]
LGTM, with a few nits.

File: lib/multiplexer/multiplexer.go [hunk 8]
```diff
@@ -386,6 +387,8 @@ func isHTTP(in []byte) bool {
 	return false
 }
 
+var proxyHelloSig = []byte(sshutils.ProxyHelloSignature)
```
Comment (espadolini) [inline lib/multiplexer/multiplexer.go:390]: nit: move near the other `fooPrefix`, maybe rename to `proxyHelloPrefix`?

File: lib/sshutils/server.go [hunk 9]
```diff
@@ -438,6 +439,14 @@ func (s *Server) HandleConnection(conn net.Conn) {
...
+				WithError(err).
+				WithFields(logrus.Fields{
+					"remote_addr": conn.RemoteAddr(),
```
Comment (espadolini) [inline lib/sshutils/server.go:447]: nit: `WithField("remote_addr", conn.RemoteAddr()).`

File: lib/multiplexer/multiplexer_test.go [hunk 10]
```diff
@@ -721,6 +731,70 @@ func TestMux(t *testing.T) {
...
+		)
+		require.NoError(t, err)
+		go srv.Serve(mux.SSH())
+		defer srv.Close()
+
+		// Manually create client conn so we can inject the ProxyHelloSignature
+		conn, err := net.DialTimeout("tcp", listener.Addr().String(), time.Second)
+		require.NoError(t, err)
+		hp := &apisshutils.HandshakePayload{
+			ClientAddr: "6.6.6.6",
```
Comment (espadolini) [inline lib/multiplexer/multiplexer_test.go:772]: Should we check that the correct IP is passed along correctly?

File: lib/service/service.go [hunk 11]
```diff
@@ -2952,10 +2952,30 @@ func (process *TeleportProcess) setupProxyListeners(networkingConfig types.Clust
...
+				Listener:            l,
+				EnableProxyProtocol: true,
+				ID:                  teleport.Component(teleport.ComponentProxy, "ssh"),
+			})
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			listeners.ssh = mux.SSH()
+			go func() {
+				if err := mux.Serve(); err != nil {
```
Comment (espadolini) [inline lib/service/service.go:2972]: Here and below we might want to do `err != nil && !utils.IsOKNetworkError(err)`.

**Review by espadolini — APPROVED** [review]

File: lib/service/service.go (hunk 11 above)
Comment (strideynet) [reply lib/service/service.go]: 3f97ad2b3

File: lib/sshutils/server.go (hunk 9 above)
Comment (strideynet) [reply lib/sshutils/server.go]: 3f97ad2b3

File: lib/multiplexer/multiplexer.go (hunk 8 above)
Comment (strideynet) [reply lib/multiplexer/multiplexer.go]: 3f97ad2b3

File: lib/multiplexer/multiplexer_test.go (hunk 10 above)
Comment (strideynet) [reply lib/multiplexer/multiplexer_test.go]: [1dd218b](https://github.com/gravitational/teleport/pull/14712/commits/1dd218b44407b6e9d4c1b2cebecc08f1361e08f8)

**Pull request merged**